web/ui/react-app/src/DataTable.test.tsx

```diff
@@ -0,0 +1,271 @@
+import * as React from 'react';
+import { shallow } from 'enzyme';
+import DataTable, { QueryResult } from './DataTable';
+import { Alert, Table } from 'reactstrap';
+import SeriesName from './SeriesName';
+
+describe('DataTable', () => {
+  describe('when data is null', () => {
+    it('renders an alert', () => {
+      const table = shallow(<DataTable data={null} />);
+      const alert = table.find(Alert);
+      expect(Object.keys(alert.props())).toHaveLength(7);
+      expect(alert.prop('color')).toEqual('light');
+      expect(alert.prop('children')).toEqual('No data queried yet');
+    });
+  });
+
+  describe('when data.result is empty', () => {
+    it('renders an alert', () => {
+      const dataTableProps: QueryResult = {
+        data: {
+          resultType: 'vector',
+          result: [],
+        },
+      };
+      const table = shallow(<DataTable {...dataTableProps} />);
+      const alert = table.find(Alert);
+      expect(Object.keys(alert.props())).toHaveLength(7);
+      expect(alert.prop('color')).toEqual('secondary');
+      expect(alert.prop('children')).toEqual('Empty query result');
+    });
+  });
+
+  describe('when resultType is a vector with values', () => {
+    const dataTableProps: QueryResult = {
+      data: {
+        resultType: 'vector',
+        result: [
+          {
+            metric: {
+              __name__: 'metric_name_1',
+              label1: 'value_1',
+              labeln: 'value_n',
+            },
+            value: [1572098246.599, '0'],
+          },
+          {
+            metric: {
+              __name__: 'metric_name_2',
+              label1: 'value_1',
+              labeln: 'value_n',
+            },
+            value: [1572098246.599, '1'],
+          },
+        ],
+      },
+    };
+    const dataTable = shallow(<DataTable {...dataTableProps} />);
+
+    it('renders a table', () => {
+      const table = dataTable.find(Table);
+      expect(table.prop('hover')).toBe(true);
+      expect(table.prop('size')).toEqual('sm');
```

Comment:
This is a really good question. A few things:

- Testing configuration: Sometimes people don't want to test configuration; sometimes they do. I would defer to your opinion and the prometheus authors' opinions on testing configuration to generally inform how we want to test components. I'm happy to take this kind of thing out.

- Derived Props: I'm used to testing style props like this  because as a react app grows, or as components get reused, generally the props of inner components become more variable, i.e., derived on the input props of the containing component. In those cases, I think it's probably a good idea to write tests, as they describe how props flow through your component tree. I may have been overzealous in this case, though! 

- React testing practices: There is a lot of active dev in the javascript/react community. There are a variety of opinions on how to test.
  - Another maybe less brittle way to test things like this is via [snapshot testing](https://jestjs.io/docs/en/snapshot-testing), which allows you to dynamically create test assertions. You can then approve or reject new snapshots when you make changes to your components.
  - Maybe a better approach that I just learned about is using [react-testing-library](https://testing-library.com/docs/react-testing-library/intro), which renders components into the DOM on a test run. That way, your tests can focus on how the user is experiencing your component (including styles) in the DOM rather than the implementation details of which component is receiving which props. 

Again, I'm happy to change things up and go in a different direction. Let me know! Or we can figure it what we like best and converge on best practices over time.

Reply:
Thanks for the thoughts! Yeah, I actually watched a video (https://www.youtube.com/watch?v=Eakp29J38YA) a couple of days ago that showed Snapshot tests as an option, and I frequently thought here "hmm, couldn't these details more easily be caught in a snapshot test rather than manually duplicating a lot of the expected properties in tests".

I think I'd be interested in exploring snapshot tests in the future, but I think at the moment what you have is fine.

Reply:
Generally is good practice to test whether props are passed properly down to components children and to the component itself of course.

Reply:
Sounds great! Happy to do that